RFC: Restricts parameter type for randn & randexp to Float types for #17608.

NHDaly opened the pull request:
Fixes `base/random.jl` and adds tests to `tests/random.jl` to restrict parametrized type parameter to one of `Float16,Float32,Float64`.

These functions already worked only for those types, but erroneously allowed parameters outside those three, with unintended behavior (crashing and infinite-looping). Now, this throws a `MethodError` if provided a non-float type-parameter.

NHDaly committed:
Fixes infinite recursion in randn for non-floats; fixes #17608.

The functions already only supported Float types
(Float16,Float32,Float64), but could be called with non-float types.
Calling with Non-Float types, e.g. `Int`, could cause an infinite loop.

Also adds tests verifying that randn throws a MethodError for types
besides Float types.

tkelman commented:
Very nice, thank you! cc @rfourquet
Same issue for BigFloat, right? May as well throw that into the tests?

NHDaly commented:
Done. Although, would we want to consider just adding `BigFloat` to the list of supported types for `randn` and `randexp`? I.e., adding it to the `Floats` union?

I can leave that for another PR after this bug-fix though. Should I leave a TODO somewhere for it?

tkelman commented:
I imagine there's code somewhere in MPFR for those, but I don't know whether we have the ccall wrappers here yet. I think a TODO as an issue, a code/test comment, or an early WIP PR would be fine.

NHDaly commented:
I've added code TODOs. Can you verify that I've formatted them in a reasonable way? (I'm new to JuliaLang contributing and don't know your conventions). :)

(Should I have written `# TODO(17627):` instead of `# TODO(#17627):`?)

tkelman commented:
3c91c9287a3a7215095f12b9f610a54710e11444 looks fine to me, we're not super uniform in formatting of todo's and fixme's

NHDaly commented:
S.g. Thanks for the review! :)

rfourquet commented on test/random.jl:
```diff
@@ -343,6 +343,23 @@ for rng in ([], [MersenneTwister()], [RandomDevice()])
+
+        @test_throws MethodError r(rng..., Number, (2,3))
+        @test_throws MethodError r(rng..., Any, 1)
+    end
```
I would add a second loop for concision: 
```
for T in [Int, Int32, String, AbstractFloat, BigFloat, Number, Any]
    @test_throws MethodError r(rng..., T)
    @test_throws MethodError r(rng..., T, 1)
    @test_throws MethodError r(rng..., T, (2, 3))
end
```
(and, why not, a loop over the dimensions: `for d in ([], [1], [(2, 3)])`).

rfourquet commented on base/random.jl:
```diff
@@ -1203,7 +1203,7 @@ let Floats = Union{Float16,Float32,Float64}
 
             # filling arrays
-            function $randfun!{T}(rng::AbstractRNG, A::AbstractArray{T})
+            function $randfun!{T<:$Floats}(rng::AbstractRNG, A::AbstractArray{T})
```
Would it work to restrict less tightly with `T<:Float` ? We can imagine that a custom RNG implements e.g. `randn(customRNG, BigFloat)`, and then wants to use the fallback `randn!(customRNG, BigFloat[1.0])`.

stevengj replied:
I don't like this either, because it will screw up `randn!` for `Complex` arrays.

I think the existing definition is the right one: `randn!` for an array calls `randn` for each element.  If the latter isn't defined, it throws a `MethodError`.

rfourquet commented:
Another possible fix would be to write a signature similar to what I used [there](https://github.com/JuliaLang/julia/blob/9fbe3b58d3352ff4c828e5c28b0924e9fd33b1bf/base/random.jl#L232) to fix a similar problem, but I guess it makes sense to restrict those `randn` functions to `AbstractFloat` anyway (cf. my inline comment for using `AbstractFloat` instead of `$Floats`). 
Also you may squash the commits when done.
Otherwise LGTM.

stevengj commented on base/random.jl:
```diff
@@ -1213,14 +1214,14 @@ let Floats = Union{Float16,Float32,Float64}
+            $randfun{T<:$Floats}(rng::AbstractRNG, ::Type{T}, dims::Dims)       = $randfun!(rng, Array{T}(dims))
+            $randfun{T<:$Floats}(rng::AbstractRNG, ::Type{T}, dims::Integer...) = $randfun!(rng, Array{T}(dims...))
+            $randfun{T<:$Floats}(                  ::Type{T}, dims::Dims)       = $randfun(GLOBAL_RNG, T, dims)
+            $randfun{T<:$Floats}(                  ::Type{T}, dims::Integer...) = $randfun(GLOBAL_RNG, T, dims...)
```
I don't like these restrictions either, for the same reason as above.  It makes it significantly harder to define e.g. `randn` for new types.

stevengj commented:
It seems like a better solution would be much simpler.  Just change:
```jl
$randfun{T}(rng::AbstractRNG, ::Type{T}, dims::Integer...) = $randfun!(rng, Array{T}(dims...))
```
to
```jl
$randfun{T}(rng::AbstractRNG, ::Type{T}, dim0::Integer, dims::Integer...) = $randfun!(rng, Array{T}(dim0, dims...))
```
That will remove the circular fallback definition for `$randfun{T}(rng::AbstractRNG, ::Type{T})`.

rfourquet commented:
Ah of course, I had already forgotten about `randn` for `Complex`. So I concur with your proposed solution (identical to what I linked to above).

NHDaly commented:
Hi, yeah that makes a ton of sense. This is a great way to break the recursion, thanks. I wasn't sure what was the best way, so I just tried the first one that came to mind. :) Thanks!

I'll try this tomorrow and send the PR back to you. Sorry for the delay; i don't have my machine with me. Thanks!

stevengj commented:
@NHDaly, note that you should just push a new commit to this PR, rather than opening a brand-new PR.

NHDaly commented:
@stevengj  Yep, thanks! I've synced, rebased, and squashed. 

This solution definitely looks better, and the tests still pass on my client. Lmk what you think! Thanks

stevengj commented on base/random.jl:
```diff
@@ -1213,14 +1213,14 @@ let Floats = Union{Float16,Float32,Float64}
-            $randfun(                                dims::Dims)       = $randfun(GLOBAL_RNG, Float64, dims)
-            $randfun(                                dims::Integer...) = $randfun(GLOBAL_RNG, Float64, dims...)
+            $randfun{T}(rng::AbstractRNG, ::Type{T}, dims::Dims                     ) = $randfun!(rng, Array{T}(dims))
+            $randfun{T}(rng::AbstractRNG, ::Type{T}, dim0::Integer, dims::Integer...) = $randfun!(rng, Array{T}(dim0, dims...))
```
Probably `dim1` would be more Julian (1-based).

NHDaly replied:
Done. Thanks!

stevengj commented on base/random.jl:
```diff
@@ -1213,14 +1213,14 @@ let Floats = Union{Float16,Float32,Float64}
+            $randfun{T}(rng::AbstractRNG, ::Type{T}, dims::Dims                     ) = $randfun!(rng, Array{T}(dims))
+            $randfun{T}(rng::AbstractRNG, ::Type{T}, dim0::Integer, dims::Integer...) = $randfun!(rng, Array{T}(dim0, dims...))
+            $randfun{T}(                  ::Type{T}, dims::Dims                     ) = $randfun(GLOBAL_RNG, T, dims)
+            $randfun{T}(                  ::Type{T}, dim0::Integer, dims::Integer...) = $randfun(GLOBAL_RNG, T, dim0, dims...)
```
No, you shouldn't use `dim0, dims...` here, you should just juse `dims...` in the lowest-level function that takes an `rng` argument.

The reason is that you still want `randn(T)` to call `randn(GLOBAL_RNG, T)`, even if there are no dimension arguments.

NHDaly replied:
I'm not sure I agree; there is already the definition at line 1203 which covers the case of empty dims. So `randn(Float32)` will still call `randn(GLOBAL_RNG, T)`.

Leaving off this change will still break the infinite-loop, but the error message is less clear. `randn(Int)` will successfully match the definition at line 1219, and then produce the error message: `MethodError: no method matching randn(::MersenneTwister, ::Type{Int64})`, which is less clear than the current error message, `MethodError: no method matching randn(::Type{Int64})`, which is what the user wrote.

Thoughts?

stevengj replied:
Ah, I missed the definition on line 1203.  I would prefer to just delete that definition.  (If there is some performance reason to define the one-argument case separately ... splatting penalty? ... we should still define it generically for all types, not just `T<:$Floats`.)

My reason is that, to add `randn` for a new type `T`, you should just have to define one method: `randn(rng::AbstractRNG, ::Type{T})`.   You shouldn't have to define other methods that we could get just by loosening the type signature.

If the price for that is slightly more obscure method errors, that's fine.   In fact, `no method matching randn(::MersenneTwister, ::Type{Int64})` is actually *more* helpful, because it tells you exactly what method you need to implement to make everything work.

NHDaly replied:
That is a great reason; I'm sold. Thanks.

So, for this PR, I'm gonna change it to work for all types, as you suggested. Maybe it makes sense to just delete that line, but I'll leave that up to future PRs?

stevengj commented on base/random.jl:
```diff
@@ -1213,14 +1213,14 @@ let Floats = Union{Float16,Float32,Float64}
+            $randfun(   rng::AbstractRNG,            dims::Dims                     ) = $randfun(rng, Float64, dims)
+            $randfun(   rng::AbstractRNG,            dim0::Integer, dims::Integer...) = $randfun(rng, Float64, dim0, dims...)
+            $randfun(                                dims::Dims                     ) = $randfun(GLOBAL_RNG, Float64, dims)
+            $randfun(                                dim0::Integer, dims::Integer...) = $randfun(GLOBAL_RNG, Float64, dim0, dims...)
```
Same here, revert this part of the patch ... the patch to `base/random.jl` should need to change only one line.

Commit:
s/dim0/dim1/

NHDaly commented:
I've left some inline comments. PTAL. thanks! :)

stevengj commented:
As noted above, I would prefer it if, to add `randn` for a new type, you should only have to define `randn(rng::AbstractRNG, ::Type{T})`.  This means making the one-argument `randn` work for any type, not just `Floats`.

NHDaly committed:
removed other dimensions, fixed randn(T), added comment.

NHDaly commented:
:) Okay I'm sold! I agree, you should only have to define `randn(rng::AbstractRNG, ::Type{T})` for your type. So I've made the one-arg version work for any type. Thanks!

Let me know if this looks okay and I'll squash the commits.

stevengj commented:
LGTM.

stevengj commented:
(No need to squash the commits, the Github merge button can do that for us.)

NHDaly commented:
> (No need to squash the commits, the Github merge button can do that for us.)

:O Woahh fancy! :) cool thanks!
